joplin/api/schema.py

```diff
@@ -340,22 +341,115 @@ class Meta:
+        for topic_collection in topic_collections:
+            topic_collection_global_id = graphene.Node.to_global_id('TopicCollectionNode', topic_collection.id)
+            site_structure.append({'url' :f'/{topic_collection.theme.slug}/{topic_collection.slug}/', 'type': 'topic collection', 'id': topic_collection_global_id});
+
+        topics = TopicPage.objects.all()
+        for topic in topics:
+            topic_global_id = graphene.Node.to_global_id('TopicNode', topic.id)
+            topic_tcs = topic.topiccollections.all()
+            for tc in topic_tcs:
+                topic_tc_global_id = graphene.Node.to_global_id('TopicCollectionNode', tc.topiccollection.id)
+                site_structure.append({'url' :f'/{tc.topiccollection.theme.slug}/{tc.topiccollection.slug}/{topic.slug}/', 'type': 'topic', 'id': topic_global_id, 'parent_topic_collection': topic_tc_global_id})
+
+        departments = DepartmentPage.objects.all()
+        for department in departments:
+            department_global_id = graphene.Node.to_global_id('DepartmentNode', department.id)
+            site_structure.append({'url' :f'/{department.slug}/', 'type': 'department', 'id': department_global_id})
+
+        service_pages = ServicePage.objects.all()
+        for service_page in service_pages:
+            service_page_global_id = graphene.Node.to_global_id('ServicePageNode', service_page.id)
+
+            if service_page.coa_global:
+                site_structure.append({'url' :f'/{service_page.slug}/', 'type': 'service page', 'id': service_page_global_id})
+
+            service_page_departments = service_page.related_departments.all()
+            for service_page_department in service_page_departments:
+                service_page_department_global_id = graphene.Node.to_global_id('DepartmentNode', service_page_department.related_department.id)
+                site_structure.append({'url' :f'/{service_page_department.related_department.slug}/{service_page.slug}/', 'type': 'service page', 'id': service_page_global_id, 'parent_department': service_page_department_global_id})
+
+            service_page_topics = service_page.topics.all()
+            for service_page_topic in service_page_topics:
+                service_page_topic_global_id = graphene.Node.to_global_id('TopicNode', service_page_topic.topic.id)
+                service_page_topic_tcs = service_page_topic.topic.topiccollections.all()
+                for tc in service_page_topic_tcs:
+                    service_page_topic_tc_global_id = graphene.Node.to_global_id('TopicCollectionNode', tc.topiccollection.id)
+                    site_structure.append({'url' :f'/{tc.topiccollection.theme.slug}/{tc.topiccollection.slug}/{service_page_topic.topic.slug}/{service_page.slug}/', 'type': 'service page', 'id': service_page_global_id, 'parent_topic': service_page_topic_global_id, 'grandparent_topic_collection': service_page_topic_tc_global_id})
```

Comment:
this gets a little nested but it's the (theme/tc/topic/page) urls as well as the ids for the parent topic and grandparent topic collection.